qualtran/_infra/bloq_example.py

```diff
@@ -0,0 +1,92 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from typing import Callable, Optional, Type
+
+from attrs import field, frozen
+
+from .bloq import Bloq
+
+
+@frozen
+class BloqExample:
+    """An instantiation of a bloq and its metadata.
+
+    In particular, this class wraps a callable that returns a bloq instantiation with
+    explicit attribute values.
+
+    Consider using the decorator `@bloq_example` to construct a `BloqExample` from a function.
+
+    Args:
+        func: The function that returns the bloq instantiation. Calling the `BloqExample` instance
+            will call this function.
+        name: A name for the bloq instantiation.
+        bloq_cls: The `Bloq` class that this instantiation is an instance of.
+        generalizer: Passed to `get_bloq_counts_graph` calls for bloq-counts equivalence checking.
+    """
+
+    _func: Callable[[], Bloq] = field(repr=False, hash=False)
```

Comment:
Can we also add a convertor so that if user directly provides a `bloq` then we can wrap it in a trivial lambda?  I can imagine calling `bloq_tester(BloqExample(bloq))` to make sure the current bloq passes all checks?

Reply:
Fair question, but I plan to use these functions in place of the current `xx_test._make_xx()` jupyter notebook autogen functions where we secretly extract the source code from it. It adds a little boilerplate for developers but will unify the examples that we document with the examples that we (auto)test.